controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+					s.jobRequests <- newReq
+				})
+			}
+		}
+	}()
+
+	log.Info("determining best host for job")
+	host, err := s.findBestHost(newReq.Formation, newReq.Type, newReq.HostID)
+	if err != nil {
+		log.Error("error determining best host for job", "err", err)
+		return err
+	}
+	hostID := host.ID()
+	newReq.HostID = hostID
+
+	config := jobConfig(newReq, hostID)
+	newReq.JobID = config.ID
+
+	// Provision a data volume on the host if needed.
+	if newReq.needsVolume() {
+		log.Info("provisioning volume")
+		if err := utils.ProvisionVolume(host, config); err != nil {
+			log.Error("error provisioning volume", "err", err)
+			return err
+		}
+	}
+
+	log.Info("requesting host to add job", "host.id", hostID, "job.id", config.ID)
+	if err := host.AddJob(config); err != nil {
+		log.Error("error requesting host to add job", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (s *Scheduler) stopJob(req *JobRequest) (err error) {
+	log := fnLogger("req.host.id", req.HostID, "req.job.id", req.JobID, "job.type", req.Type)
+	log.Info("stopping job")
+	defer func() {
+		if err != nil {
+			log.Error("error stopping job", "err", err)
+		}
+	}()
+
+	var job *Job
+	if req.JobID == "" {
+		formationKey := utils.FormationKey{AppID: req.AppID, ReleaseID: req.ReleaseID}
+		typJobs := s.jobs.GetStoppableJobs(formationKey, req.Type)
+		if len(typJobs) == 0 {
+			e := fmt.Sprintf("no %s jobs running", req.Type)
+			log.Error(e)
+			return errors.New(e)
+		}
+		job = typJobs[0]
+		for _, j := range typJobs {
+			if j.startedAt.After(job.startedAt) {
+				job = j
+			}
+		}
+		log.Info("selected job for termination", "job.id", job.JobID, "job.host.id", job.HostID)
+	} else {
+		var ok bool
+		job, ok = s.jobs[req.JobID]
+		if !ok {
+			e := "unknown job"
+			log.Error(e)
+			return errors.New(e)
+		}
+	}
+	job = s.handleJobStop(job)
+	if job.HostID != "" {
+		log.Info("getting host client")
+		host, err := s.Host(job.HostID)
+		if err != nil {
+			log.Error("error getting host client", "err", err)
+			s.unfollowHost(job.HostID)
+			return err
+		}
+
+		log.Info("requesting host to stop job")
+		if err := host.StopJob(job.JobID); err != nil {
+			log.Error("error requesting host to stop job", "err", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func jobConfig(req *JobRequest, hostID string) *host.Job {
+	return utils.JobConfig(req.Job.Formation.ExpandedFormation, req.Type, hostID)
+}
+
+func (s *Scheduler) findBestHost(formation *Formation, typ, hostID string) (utils.HostClient, error) {
+	log := fnLogger("app.id", formation.App.ID, "release.id", formation.Release.ID, "job.type", typ)
+	log.Info("getting host list")
+	hosts, err := s.getHosts()
+	if err != nil {
+		log.Error("error getting host list", "err", err)
+		return nil, err
+	}
+	if len(hosts) == 0 {
+		e := "no hosts found"
+		log.Error(e)
+		return nil, errors.New(e)
+	}
+
+	if hostID != "" {
+		log.Info("using explicit host", "host.id", hostID)
+		return s.Host(hostID)
+	}
```

Comment:
Under what circumstances is `startJob` provided with an explicit host ID? Should we not be picking a host each time?